### PR TITLE
Add tests to existing trimperson api

### PR DESCRIPTION
**Pull request by ebuckley**
Add tests for the TrimPerson api, pulls in [mocha](http://mochajs.org/) [chai](http://chaijs.com/api/bdd/) and [nock](https://www.npmjs.com/package/nock)

If you are checking this out, you will need to do an `npm install`

run tests in single-run mode by using `npm run test`
run tests in watch mode by using `npm run test:watch`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
   "description": "TRIM API wrapper",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "./node_modules/mocha/bin/mocha specs.js",
```
Comment (ebuckley) [inline package.json]: We should move tests out into a tests folder or have a convention for test files that mocha picks up. I have seen projects where the test runner simply looks for a glob such as `**/*.spec.js` which allows your specs to be local to the modules they test.

For now I have just included a single `specs.js` for holding all tests.

File: specs.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+          subContainers: []
+        })
+
+      trim.getContainer(fakeID, function (err, data) {
+        expect(err).to.be.an('null')
+        expect(data).not.to.be.a('undefined')
+
+        expect(data.containerNo).to.equal('someid')
+        expect(data.subContainers).to.be.a('array')
+        getContainerMock.done() // throws an error if no request was recorded
```
Comment (ebuckley) [inline specs.js]: calling `.done()` will assert that a request with matching query params, body and header has been made.

File: specs.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      expect(trim.getDocument).to.be.a('function')
+    })
+
+    it('should getDocument succesfully', function (done) {
+      var documentId = 'someIDthatPointsToADoucment'
+      var getDocumentMock = nock(mockURL)
+        .get('/get')
+        .query({id: documentId, securityToken: mockToken})
+        .reply(200, {
+          msg: 'response' // TODO what shape object does getDocument actually return
```
Comment (ebuckley) [inline specs.js]: I wasn't sure what the expected shape of the getDocument response is so I faked it, the production code doesn't check the shape anyway so returning the correct shape is a moot point.

Comment (pnw) [inline specs.js]: getDocument responds with the actual file bytecode, e.g. a PDF file. It's not a json

**Comment by pnw** [comment]
Awesome! One question: these are testing against the callback style methods. What do we have to change to make them promise-compatible?

**Comment by ebuckley** [comment]
We can add specs that look essentially the same but change the invocation from:
```js
trim.createRecord('newTitle', 'containerID', 'extensiontype', 'fileData', function (err, data) {
        expect(err).to.be.an('null')
        expect(data).not.to.be.an('undefined')

        expect(data).to.be.a('string')
        expect(data).to.equal('123456')
        createRecordMock.done()
        done()
      })
```

To: 
```js
trim.createRecord('newTitle', 'containerID', 'extensiontype', 'fileData')
.then( function (data) {
        expect(data).not.to.be.an('undefined')

        expect(data).to.be.a('string')
        expect(data).to.equal('123456')
        createRecordMock.done()
        done()
}, function (err) { 
        throw err // assuming that a failure case causes this spec to fail
        done()
})
```

File: package.json (hunk 1 above)
Comment (digitalsadhu) [inline package.json]: Yup that's the way we usually do it.

File: package.json [hunk 4]
```diff
@@ -4,7 +4,8 @@
...
+    "test:watch": "./node_modules/nodemon/bin/nodemon.js --exec \"npm run test\" specs.js"
```
Comment (digitalsadhu) [inline package.json]: Have you seen testem? We've used it quite a bit to rerun tests when files change.

File: specs.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        .get('/GetContainer')
+        .query({trimid: fakeID, securityToken: mockToken})
+        .reply(200, {
+          containerNo: 'someid',
+          subContainers: []
+        })
+
+      trim.getContainer(fakeID, function (err, data) {
+        expect(err).to.be.an('null')
+        expect(data).not.to.be.a('undefined')
```
Comment (markstuart) [inline specs.js]: Is there a nicer way to test for undefined? This seems a bit... weird

File: package.json (hunk 4 above)
Comment (digitalsadhu) [inline package.json]: See: https://github.com/mediasuitenz/mappy

For our most up to date test setup including testem

File: specs.js (hunk 5 above)
Comment (markstuart) [inline specs.js]: oh, I see in the Chia expect API docs we can use `.undefined`:
```js
expect(data).not.to.be.undefined
```

Comment (digitalsadhu) [inline specs.js]: Doesn't chai expect have a not.to.be.undefined?

Comment (digitalsadhu) [inline specs.js]: Oooh snappity

File: specs.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    })
+  })
+
+  describe('when using the getDocument function', function () {
+    it('should exist on the trim instance', function () {
+      expect(trim.getDocument).to.be.a('function')
+    })
+
+    it('should getDocument succesfully', function (done) {
+      var documentId = 'someIDthatPointsToADoucment'
```
Comment (markstuart) [inline specs.js]: What is a `Doucment`? Is this some Douche related thing? :smiling_imp:

File: specs.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      var createRecordMock = nock(mockURL)
+        .filteringRequestBody(function (bodyt) {
+          return '*'
+        })
+        .post('/AddRecordToTrim', '*')
+        .query({ securityToken: mockToken })
+        .reply(500, { msg: 'it failed wtfomg' }) // determine shape of response
+
+      trim.createRecord('newTitle', 'containerID', 'extensiontype', 'fileData', function (err, data) {
+        expect(err).not.to.be.an('null')
```
Comment (markstuart) [inline specs.js]: This could be `expect(err).not.to.be.null` or `expect(err).to.exist` which might be nicer. `.exist` checks for `!null` or `!undefined` apparently.

Comment (markstuart) [inline specs.js]: Well I guess technically it's `!null && !undefined`

File: package.json (hunk 4 above)
Comment (markstuart) [inline package.json]: Yeah, Richie and I did quite a bit of iteration on test style in that project. Do you remember which tests were the better ones for Ersin to look at in there @digitalsadhu? From memory there are some older ones that we were... less proud of.

File: specs.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
+var describe = require('mocha').describe
+var it = require('mocha').it
+var expect = require('chai').expect
+var nock = require('nock')
+
+var mockURL = 'http://localhost'
+var mockToken = 'DEADBEEF'
```
Comment (markstuart) [inline specs.js]: :laughing:

File: package.json (hunk 4 above)
Comment (digitalsadhu) [inline package.json]: Maybe something like: https://github.com/mediasuitenz/mappy/blob/master/test/key.spec.js

We were using Mocha GWT and arrow functions to clean up the test cases.
I did a follow up with Justin at Test Double and he recommended we drop the text in the Givens, Whens and Thens though as he felt it creates noise so an ideal test looks something like:

```js
describe('setting the maps title', () => {
  Given(() => title = 'My title')
  When(() => key.title = title)
  Then(() => key.title.should.equal(title))
})
```

Comment (digitalsadhu) [inline package.json]: And if anyone is interested in using GWT syntax, there are now 2 mocha libs to choose from. The one we use in mappy:
https://github.com/rendro/mocha-given

And a newer more actively maintained version here:
https://github.com/TheLudd/mocha-gwt

**Comment by digitalsadhu** [comment]
Looking sharp!

**Comment by ebuckley** [comment]
Thanks for all the feedback!

Changes:
- Use Given,When, Then
- Use testem for test watching.
- Use babelify for test code

File: specs.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+describe('when using trimperson api', () => {
+  var trim
+  var mockURL
+  var mockToken
+
+  Given(() => mockURL = 'http://localhost')
+  Given(() => mockToken = 'DEADBEEF')
+
+  Given(() => {
```
Comment (digitalsadhu) [inline specs.js]: this looks like the "When" to me. The action being creating/setting up the TRIM object

File: specs.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+          expect(data).to.exist
+
+          expect(data.containerNo).to.equal('someid')
+          expect(data.subContainers).to.be.a('array')
+          getContainerMock.done() // throws an error if no request was recorded
+          done()
+        })
+      })
+    })
+  })
```
Comment (digitalsadhu) [inline specs.js]: I would expect to see a When in here too. Something like this:

```js
describe('when invoking the getContainer function', () => {
   var getContainerMock
   var fakeID = 'fakecontainerid'
   var returnedData
   var returnedError

   Given(() => {
     getContainerMock = nock(mockURL)
       .get('/GetContainer')
       .query({trimid: fakeID, securityToken: mockToken})
       .reply(200, {
         containerNo: 'someid',
         subContainers: []
       })
   })

   When((done) => trim.getContainer(fakeID, (err, data) => {
    returnedError = err
    returnedData = data
    done()
   })

   Then(() => {
     expect(returnedError).not.to.exist
     expect(returnedData).to.exist
     expect(returnedData.containerNo).to.equal('someid')
     expect(returnedData.subContainers).to.be.a('array')
     getContainerMock.done() // throws an error if no request was recorded
   })
 })
```

File: specs.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        getDocumentMock.done()
+        done()
+      })
+    })
+  })
+
+  describe('when using the createContainer function', () => {
+    var createContainerMock, folder, parentFolder, privacySetting
+    Given(() => folder = 'someFolder')
+    Given(() => parentFolder = 'someParentFolder')
```
Comment (digitalsadhu) [inline specs.js]: you should be able to use And here I think if you wish though I'm perfectly happy with multiple Given statements.

**Comment by digitalsadhu** [comment]
@ebuckley nice work!

Each test should have at least 1 Given, 1 When and 1 Then (plus any Ands or extra GWTs), otherwise lookin sharp! :+1:

**Comment by ebuckley** [comment]
I was adverse to adding `When`'s because the `When` was also the `Then`.
Would the following be idiomatic GWT?
```js
 When('it should getDocument succesfully', (done) => {
     trim.getDocument(documentId, function (err, data) {
       Then( () => {
           expect(err).not.to.exist
           expect(data).to.exist
           getDocumentMock.done()
           done()
       })
      })
    })
```

Also ready to merge this, was thinking it should happen before #4 and #3 are merged

**Comment by digitalsadhu** [comment]
@ebuckley I think you are right except that theres no need to nest, you can use the done callback to separate the blocks, see: https://github.com/searls/jasmine-given#testing-asynchronous-code
(mocha-given is a direct port of jasmine-given)
The idea is to use the done callback so you can split out when and then. The When being the action of calling the method `trim.getDocument` and the Then being the expectations that follow.

**Comment by pnw** [comment]
Getting some errors here:

1. couldn't find `mocha-babel` package after `npm install`
2. Given is not defined

**Comment by ebuckley** [comment]
Try it out now @pnw
